src/algebra/universal_enveloping_algebra.lean

```diff
@@ -0,0 +1,158 @@
+  map_mul'  := λ r s, by { change _ = quot.mk _ _, simp [mul_smul], },
+  map_zero' := by { rw zero_smul, refl, },
+  map_add'  := λ r s, by { change _ = quot.mk _ _, rw add_smul, },
+  commutes' := λ r, by { rintros ⟨a⟩, change quot.mk _ _ = quot.mk _ _, simp, },
+  smul_def' := λ r u, rfl, }
+
+instance : inhabited (universal_enveloping_algebra R L) := ⟨0⟩
+
+/-- The quotient map from the tensor algebra to the universal enveloping algebra as a morphism of
+associative algebras.
+
+This map can be shown to exist using just universal properties but it is convenient to compromise
+slightly and construct it directly as below. -/
+def mk_alg_hom : tensor_algebra R L →ₐ[R] universal_enveloping_algebra R L :=
+{ to_fun    := quot.mk (rel R L),
+  map_one'  := rfl,
+  map_mul'  := λ a b, rfl,
+  map_zero' := rfl,
+  map_add'  := λ a b, rfl,
+  commutes' := λ r, show _ = quot.mk _ _, by rw [algebra.smul_def'', mul_one], }
+
+/-- The natural Lie algebra morphism from a Lie algebra to its universal enveloping algebra.
+
+Note that the Lie algebra structure on the universal enveloping algebra arises from its associative
+algebra structure via the ring commutator. -/
+def ι : L →ₗ⁅R⁆ universal_enveloping_algebra R L :=
+{ to_fun    := quot.mk (rel R L) ∘ (tensor_algebra.ι R L),
+  map_add'  := λ x y, by simpa only [function.comp_app, linear_map.map_add],
+  map_smul' := λ c x, show _ = quot.mk _ _, by simp,
+  map_lie   := λ x y, quot.sound (rel.lie_compat x y), }
+
+variables (A : Type u₃) [ring A] [algebra R A] (f : L →ₗ⁅R⁆ A)
+
+/-- The universal property of the universal enveloping algebra: Lie algebra morphisms into
+associative algebras lift to associative algebra morphisms from the universal enveloping algebra. -/
+def lift : universal_enveloping_algebra R L →ₐ[R] A :=
+{ to_fun    := λ a, quot.lift_on a (tensor_algebra.lift R L (f : L →ₗ[R] A))
+    (λ a b h,
+    begin
+      induction h,
+      { -- case universal_enveloping.rel.lie_compat
+        rw [← alg_hom.to_linear_map_apply, ← linear_map.comp_apply, tensor_algebra.ι_comp_lift],
+        change f _ = _, simpa [lie_ring.of_associative_ring_bracket], },
+      { -- case universal_enveloping.rel.add_compat
+        simp only [alg_hom.map_add, add_left_inj, h_ih], },
+      { -- case universal_enveloping.rel.mul_compat_left
+        simp only [alg_hom.map_mul, h_ih], },
+      { -- case universal_enveloping.rel.mul_compat_right
+        simp only [alg_hom.map_mul, h_ih], },
+    end),
+  map_one'  := show algebra_map _ _ _ = _, by rw ring_hom.map_one,
+  map_mul'  := by { rintros ⟨⟩ ⟨⟩, change tensor_algebra.lift R L ↑f _ = _, rw alg_hom.map_mul, },
+  map_zero' := show tensor_algebra.lift R L ↑f 0 = (0 : A), by rw alg_hom.map_zero,
+  map_add'  := by { rintros ⟨⟩ ⟨⟩, change tensor_algebra.lift R L ↑f _ = _, rw alg_hom.map_add, },
+  commutes' := λ r, show tensor_algebra.lift R L ↑f _ = _, by simp [algebra.smul_def''], }
+
+lemma ι_comp_lift : (lift R L A f) ∘ (ι R L) = f :=
+by { ext, refl, }
```

Comment:
A @[simp] lemma variant, applied to an argument?

Reply:
Done.